File: ml-benchmark/synduce/constraints/empty_right_subtree/sum.f
Inductive Tree = nil Unit | leaf Int | node {Int, Tree, Tree};

size = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    nil _ -> 0
  | leaf _ -> 1
  | node {_, l, r} -> + 1 (+ (f l) (f r))
  end
);

empty_right = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    node {_, l, r} -> and (== 0 (size r)) (f l)
  | _ -> true
  end
);

spec = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    nil _ -> 0
  | leaf a -> a
  | node {a, l, r} -> + a (+ (f l) (f r))
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | leaf a -> leaf a
  | node {a, l, r} -> node {a, f l, r} /* Avoid the recursion on the right branch.*/
  end
);

main = \t: Tree. if empty_right t then spec (target t) else 0;

File: ml-benchmark/synduce/constraints/empty_right_subtree/sum.ml
type tree = Nil of unit | Leaf of int | Node of int * tree * tree

let rec size t =
  match t with
  | Nil _ -> 0
  | Leaf _ -> 1
  | Node (_, l, r) -> 1 + (size l + size r)

let rec empty_right t =
  match t with
  | Node (_, l, r) -> (0 == size r) && empty_right l
  | _ -> true

let rec spec t =
  match t with
  | Nil _ -> 0
  | Leaf a -> a
  | Node (a, l, r) -> a + (spec l + spec r)

val target: tree -> tree compress
let rec target t =
  match t with
  | Nil _ -> Nil ()
  | Leaf a -> Leaf a
  | Node (a, l, r) -> Node (a, target l, r)

let program t =
  if empty_right t then spec (target t) else 0